vignettes/concepts_conventions.Rmd

```diff
@@ -0,0 +1,285 @@
+
+## Common pitfalls 
+
+Expressions are very powerful, but this can also lead to misunderstandings about their functionality. Let's set up some dummy
+data to explore common issues that new (or experienced!) programmers may encounter when dealing with expressions.
+
+```{r, eval = TRUE, echo = TRUE}
+library(dplyr, warn.conflicts = FALSE)
+library(admiral)
+
+vs <- tribble(
+  ~USUBJID, ~VSTESTCD, ~VISIT, ~VSSTRESN, ~VSSTRESU, ~VSDTC,
+  "01-1301", "WEIGHT", "SCREENING", 82.1, "kg", "2013-08-29",
+  "01-1301", "WEIGHT", "WEEK 2", 81.19, "kg", "2013-09-15",
+  "01-1301", "WEIGHT", "WEEK 4", 82.56, "kg", "2013-09-24",
+  "01-1302", "BMI", "SCREENING", 20.1, "kg/m2", "2013-08-29",
+  "01-1302", "BMI", "WEEK 2", 20.2, "kg/m2", "2013-09-15",
+  "01-1302", "BMI", "WEEK 4", 19.9, "kg/m2", "2013-09-24"
+)
+
+dm <- tribble(
+  ~USUBJID, ~AGE,
+  "01-1301", 18
+)
+```
+
+### 1. Mistakenly passing something that isn't an expression to an argument {#pitfall1}
+
+When writing more complex `{admiral}` code it can be easy to mistakenly pass the wrong input to an argument that 
+expects an expression. For example, the code below fails because `my_expression` is not an expression - is 
+the name of an object in the global environment containing an expression.
+
+```{r, eval = TRUE, error = TRUE}
+my_expression <- expr(VSTESTCD == "WEIGHT" & VISIT == "SCREENING")
+
+derive_vars_merged(
+  dm,
+  dataset_add = select(vs, USUBJID, VSTESTCD, VISIT),
+  by_vars = exprs(USUBJID),
+  filter_add = my_expression
+)
+```
+
+To fix this code, we need to [unquote](#unquoting) `my_expression` so that the expression that it is holding is passed correctly
+to `derive_vars_merged()`:
+
+```{r, eval = TRUE, error = FALSE}
+derive_vars_merged(
+  dm,
+  dataset_add = select(vs, USUBJID, VSTESTCD, VISIT),
+  by_vars = exprs(USUBJID),
+  filter_add = !!my_expression
+)
+```
+
+### 2. Forgetting that expressions must be evaluable in the dataset
+
+In a similar vein to above, even if an actual expression *is* passed as an argument, you must make
+sure that it can be evaluated within the dataset of interest. This may seem trivial, but it is a common pitfall because
+expressions delay evaluation of code and so can delay the identification of issues. For instance, consider this example:
+
+```{r, eval = TRUE, error = TRUE}
+filter_vs_and_merge <- function(my_expression) {
+  derive_vars_merged(
+    dm,
+    dataset_add = select(vs, USUBJID, VSTESTCD, VISIT),
+    by_vars = exprs(USUBJID),
+    filter_add = !!my_expression
+  )
+}
+
+# This works
+filter_vs_and_merge(expr(VSTESTCD == "WEIGHT" & VISIT == "SCREENING"))
+
+# This fails
+filter_vs_and_merge(expr(VSTESTCD == "WEIGHT" & VISIT == "SCREENING" & VSTPT == "PREDOSE"))
+```
+
+The second call fails because hidden within the expression is a mention of `VSTPT`, which is a variable missing from `vs`.
```

Comment:
```suggestion
The second call fails because hidden within the expression is a mention of `VSTPT`, which was dropped from `vs` in `filter_vs_and_merge()`.
```